Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs

```diff
@@ -0,0 +1,141 @@
+using System.Collections.Generic;
+using Unity.Netcode;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    public class SceneLoader : NetworkBehaviour
+    {
+        /// <summary>
+        /// This singleton is a utility handling the loading of scenes. It loads scene using the SceneManager, or, on
+        /// listening servers for which scene management is enabled, using the NetworkSceneManager. It also starts and
+        /// stops the loading screen.
+        /// </summary>
+
+        [SerializeField]
+        ClientLoadingScreen m_ClientLoadingScreen;
+
+        public static SceneLoader Instance { get; private set; }
+
+        public void Awake()
+        {
+            if (Instance != null && Instance != this)
+            {
+                Destroy(this.gameObject);
+            }
+            else
+            {
+                Instance = this;
+            }
+            DontDestroyOnLoad(this);
```

Comment:
I know Netcode used to have issues with DontDestroyOnLoad and scene switching. That's solved right? If you finish a game and try to start a new one, this should work fine correct?

Reply:
It does not work fine. If I start a game, then quit, then start a new one, I get an exception on the client that says: `Unhandled RPC Exception: System.InvalidCastException: Specified cast is not valid.` when joining. It comes from `StopLoadingScreenClientRpc`.

I'll try to debug this

Reply:
Pre.4 fixes this issue, but there is another one still. In short, if the host is shut down, then restarted, the NetworkObject is not re-spawned. I created a [github issue](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/1574) for it. I think we should put this PR on hold in the meantime.

Reply:
When you say restarted and the NO is not respawned, you mean host side right? Cause I wouldn't expect a client to still be connected after a shutdown.
So when starting a 2nd game in the same process, the host can't have that object present a second time? yeah that's bad :(

Reply:
It actually happens on both the host and the client. After leaving a game, if any of the two game processes try to restart a new game, the NetworkObject will not be spawned. As you intuited, it doesn't make the game unplayable, it simply doesn't activate the loading screen on the client (it does still activate them on the host, though). However, if we shutdown the session a second time, we will be stuck in the main menu with an infinite loading screen. In short, the game is playable once normally, then a second time with no loading screen on the client, but not a third time.

If, however, we have a host or client that has started hosting or joined a game, then quits, then tries to join a game that is hosted by a new process, it will give use exceptions and fail to spawn the NetworkObject, which makes the game unplayable. The reverse case (a host that has a previous session joined by a new client) does not cause this issue.

Reply:
yeah ok, I agree, let's keep this on hold then